### PR TITLE
GH-3114: Honor SpEL contract in ExpressionEvalMap

### LINKED ISSUE #3114: NULL HttpEntity in headerFunction in HTTP inbound gateway
When trying to upgrade from version 5.1.8 to 5.2.1, our HTTP inbound gateway started throwing NPEs when evaluating header functions. Our code is quite basic:
```
IntegrationFlows.from(
                        inboundGateway( gatewayPath )
                                .headerFunction(SOME_HEADER, ent -> getSomeHeaderValue(ent ) )  // ent is null
```

Seems to be related to the call `FunctionExpression.getValue(...)` which ignores `EvaluationContext.getRootObject()`

Do you need more info on this?

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-projects/spring-integration/issues/3114

The contract of SpEL with its
`getValue(EvaluationContext context, @Nullable Object rootObject)` is
that we need to deal with provided `rootObject` even if it is `null`
and don't consult with `context.getRootObject()`

* Fix `ExpressionEvalMap` to have an internal `rootExplicitlySet`
to indicate that `root` explicitly provided by consumer, even if it is null.
According this flag call respective `Expression.getValue()`
* Add `@Nullable` to methods and their arguments into `ExpressionEvalMap`
& `FunctionExpression` to honor `Expression` contracts
* Populate an `HttpEntity` explicitly into `ExpressionEvalMap` from the
`HttpRequestHandlingEndpointSupport` and `WebFluxInboundEndpoint` for
full picture

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java [hunk 1]
```diff
@@ -51,8 +51,8 @@
  *	.usingCallback(new EvaluationCallback() {
  *		Object evaluate(Expression expression) {
  *			// return some expression evaluation
- *		}
- *	})
+ *        }
+ *    })
  *	.build();
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java]: Change from tabs to spaces has messed up alignment.

File: spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java [hunk 2]
```diff
@@ -206,22 +207,35 @@ public static ExpressionEvalMapBuilder from(Map<String, ?> expressions) {
...
+		@Nullable
 		private final Class<?> returnType;
 
-		public ComponentsEvaluationCallback(EvaluationContext context, Object root, Class<?> returnType) {
+		public ComponentsEvaluationCallback(@Nullable EvaluationContext context, @Nullable Object root,
+				boolean rootExplicitlySet, @Nullable Class<?> returnType) {
+
 			this.context = context;
 			this.root = root;
+			this.rootExplicitlySet = rootExplicitlySet;
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java]: It's not clear what the purpose of this is; we still get an NPE if the root has explicitly been set to `null`.

>is that we need to deal with provided rootObject even if it is `null`

When I changed the HRHES to `.withRoot(null)` BOOM!!

**Review by garyrussell — CHANGES_REQUESTED** [review]
See comments; need a test with explicit `null` root.

File: spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java (hunk 2 above)
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java]: That's correct: it is exactly what SpEL contract does in the `getValue(EvaluationContext context, @Nullable Object rootObject)`.
If you explicitly provide a root as `null`, then it is used as is, without any consultation with `context`.

So, we fix an NPE on our level because we really have a root, but we still don't break SpEL contract when root is explicitly provided as `null`

File: spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java (hunk 1 above)
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java]: Doesn't look like, but I've changed everything into spaces. I don't think tabs are good for JavaDocs.

File: spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java (hunk 2 above)
Comment (garyrussell) [reply spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java]: So you are saying the NPE is ok,  because it's in "user" code (in this case, the lambda) so the user is responsible for a null check?

File: spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java (hunk 1 above)
Comment (garyrussell) [reply spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java]: Right, but you had mixed tabs and spaces here. The indentation is still wrong (extra space before `}`). Will fix on merge.

File: spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java (hunk 2 above)
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java]: That's right if (s)he provides a `null` explicitly, but in our case we just miss the proper root to propagate.
See all the `Expression.getValue()` with `root` - it is `@Nullable`. Therefore without checking for `null` it is really going to lead into NPE.
But in our case we provide all the stuff for expression environment, therefore it is safe not check because we guarantee now a non-null root.
See changes in those HTTP components.

**Comment by garyrussell** [comment]
Merged as 516ecbcf8b
